src/components/SummaryCard/SummaryCardContainer/SummaryCardContainer.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * @file Summary card container.
+ * @copyright IBM Security 2019
+ */
+
+import { arrayOf, shape, string } from 'prop-types';
+import React, { useState } from 'react';
+
+import { appendComponentNamespace } from '../../../globals/namespace';
+
+import { namespace as summaryCardNamespace } from '../SummaryCard';
+import { namespace as summaryCardSelectNamespace } from '../SummaryCardSelect/SummaryCardSelect';
+
+const getSummaryCard = (summaryCards, selectedId) =>
+  summaryCards.filter(({ id }) => id === selectedId);
+
+const resetSelectedSummaryCards = state => state([]);
+
+function SummaryCardContainer({ render, summaryCards }) {
+  const [
+    selectedSummaryCards,
+    setSelectedSummaryCards,
+  ] = resetSelectedSummaryCards(useState);
+
+  const { length: totalSelected } = selectedSummaryCards;
+
+  function getBatchActionProps({ ...props }) {
+    return {
+      ...props,
+      onCancel: () => resetSelectedSummaryCards(setSelectedSummaryCards),
+      shouldShowBatchActions: totalSelected > 0,
+      totalSelected,
+    };
+  }
+
+  function isSummaryCardSelected(selectedId) {
+    return getSummaryCard(selectedSummaryCards, selectedId).length > 0;
+  }
+
+  function onSelect(selectedId) {
+    return setSelectedSummaryCards(
+      isSummaryCardSelected(selectedId)
+        ? selectedSummaryCards.filter(({ id }) => id !== selectedId)
+        : selectedSummaryCards.concat(getSummaryCard(summaryCards, selectedId))
+    );
+  }
+
+  function getSelectionProps({ id, ...props }) {
+    return {
+      ...props,
+      checked: isSummaryCardSelected(id),
+      id: appendComponentNamespace(summaryCardSelectNamespace, id),
+      onChange: () => onSelect(id),
+    };
+  }
+
+  const renderProps = {
+    getBatchActionProps,
+    getSelectionProps,
+    selectedSummaryCards,
+    summaryCards,
+  };
+
+  return (
+    <div
+      className={appendComponentNamespace(summaryCardNamespace, 'container')}
+    >
+      {render(renderProps)}
+    </div>
+  );
+}
+
+SummaryCardContainer.propTypes = {
+  /** Provide a list of all the summary cards to render in the container */
+  summaryCards: arrayOf(
+    shape({
+      id: string.isRequired,
+    })
+  ).isRequired,
+};
```

Comment:
Is there any way avoid this situation? This feels like a divergence from the prev structure of the component, where we were avoiding props like this?

This feels like a mixture of establishing the API and also enforcing how the user adds `SummaryCard`s to the container. Could they just add `children` in their own way?

Reply:
As far as I understand it, we'll always need an array of objects containing at least an ID (any other custom properties defined by the consumer are passed straight through), so we can track what cards have been selected and manage the logic of the batch actions, similarly to how Carbon handles this for the `DataTable`.

I was searching for examples of how consumers might currently display lists of cards, but because I couldn't find any currently in-product, the closest example we have is the `QueryCard`, which this is based off - https://github.ibm.com/investigate/investigate/blob/develop/src/components/queryCardList/QueryCardList.js#L45

My assumption is that the data for these are always going to be an array that's fetched from somewhere, and then subsequently looped through to be rendered. The renderProp should still enable teams to render the cards as they wish, the only requirement being an ID on our end for state management.

I feel like this is acceptable, but I'm open to an alternative solution if it satisfies those requirements. Would you have an example of how this might work?

Reply:
Thank you for the explanation @SimonFinney! I see what you mean now. 😅